### PR TITLE
Refactored code, new way to generate phonological groups

### PR DESCRIPTION
**Pull request by alvp**

### CODE REVIEW

File: src/rantanplan/core.py [hunk 1]
```diff
@@ -360,51 +463,86 @@ def spacy_tag_to_dict(tag):
...
+                     or (pos == "PRON" and tag.get("PronType", None) in (
+                                "Prs", "Ind"))
+                     or (pos == "DET" and tag.get("PronType", None) == "Ind")
+                     or (pos in (
+                                "ADJ",
+                                "DET" and tag.get("Poss", None) == "Yes"))
+                     or (pos in ("PRON", "DET") and tag.get("PronType",
+                                                            None) in (
+                                 "Exc", "Int", "Dem"))
+                     or "".join(word).lower() in POSSESSIVE_PRON)):
```
Comment (versae) [inline src/rantanplan/core.py]: I've tried to make this a bit more readable:
```        first_monosyllable = syllable_list[0].lower()
        if ((first_monosyllable not in UNSTRESSED_UNACCENTED_MONOSYLLABLES) 
                and (first_monosyllable in STRESSED_UNACCENTED_MONOSYLLABLES
                     or pos not in ("SCONJ", "CCONJ", "DET", "PRON", "ADP")
                     or (pos == "PRON" and tag.get("Case") == "Nom")
                     or (pos == "DET" and tag.get("Definite") in (
                         "Dem", "Ind"))
                     or pos in ("PROPN", "NUM", "NOUN", "VERB", "AUX", "ADV")
                     or (pos == "ADJ" and tag.get("Poss", None) != "Yes")
                     or (pos == "PRON"
                            and tag.get("PronType", None) in ("Prs", "Ind"))
                     or (pos == "DET" and tag.get("PronType", None) == "Ind")
                     or (pos in ("ADJ", "DET"
                            and tag.get("Poss", None) == "Yes"))
                     or (pos in ("PRON", "DET")
                            and tag.get("PronType", None) in (
                                "Exc", "Int", "Dem"))
                     or "".join(word).lower() in POSSESSIVE_PRON)):
```
Let me know if it works.

File: src/rantanplan/core.py [hunk 2]
```diff
@@ -360,51 +463,86 @@ def spacy_tag_to_dict(tag):
...
+    elif pos in ("INTJ", "PROPN", "NUM", "NOUN", "VERB", "AUX", "ADV") \
+            or (pos == "ADJ" and
+                "".join(word).lower() not in POSSESSIVE_PRON_UNSTRESSED) \
+            or (pos == "PRON" and tag.get("PronType", None) in ("Prs", "Ind")) \
+            or (pos == "DET" and tag.get("PronType", None) in ("Dem", "Ind")) \
+            or (pos == "DET" and tag.get("Definite", None) == "Ind") \
+            or (pos == "PRON" and tag.get("Poss", None) == "Yes") \
+            or (pos in ("PRON", "DET") and tag.get("PronType", None) in (
+            "Exc", "Int", "Dem")) \
+            or ("".join(word).lower() in POSSESSIVE_PRON):
```
Comment (versae) [inline src/rantanplan/core.py]: Try using parenthesis.

File: src/rantanplan/core.py [hunk 3]
```diff
@@ -360,51 +463,86 @@ def spacy_tag_to_dict(tag):
...
         # Sinaeresis
         first_syllable = syllable_list[index - 1]
         second_syllable = syllable
         if first_syllable and second_syllable and (
-                (first_syllable[-1] in STRONG_VOWELS and second_syllable[0] in STRONG_VOWELS)
-                or (first_syllable[-1] in WEAK_VOWELS and second_syllable[0] in STRONG_VOWELS)
-                or (first_syllable[-1] in STRONG_VOWELS and second_syllable[0] in WEAK_VOWELS)):
-            out_syllable_list[index-1].update({'has_sinaeresis': True})
+                (first_syllable[-1] in STRONG_VOWELS and second_syllable[
+                    0] in STRONG_VOWELS)
```
Comment (versae) [inline src/rantanplan/core.py]: Try breaking the line in the `and` instead.

File: src/rantanplan/core.py [hunk 4]
```diff
@@ -491,4 +649,86 @@ def get_scansion(text, rhyme_analysis=False, rhythm_format="pattern"):
...
+        if line["rhythm"]["length"] < 11:
+            for candidate in generate_phonological_groups(raw_tokens[idx]):
+                rhythm = get_rhythmical_pattern(candidate, rhythm_format)
+                if rhythm["length"] == 11 and rhythm["stress"][-1] == "-":
+                    line.update({
+                       "phonological_groups": candidate,
+                       "rhythm": rhythm,
+                    })
+                    break
+    """
```
Comment (versae) [inline src/rantanplan/core.py]: Can we put it all together?
```
        for idx, line in enumerate(lines):
            structure_length = STRUCTURES_LENGTH.get(line["structure"], None)
            if structure_length is not None:
                if line["rhythm"]["length"] < structure_length[idx]:
                    candidates = generate_phonological_groups(raw_tokens[idx])
                    for candidate in candidates:
                        rhythm = get_rhythmical_pattern(
                            candidate, rhythm_format)
                        if rhythm["length"] == 11:
                            line.update({
                            "phonological_groups": candidate,
                            "rhythm": rhythm,
                            })
                            break
```

File: src/rantanplan/core.py [hunk 5]
```diff
@@ -491,4 +649,86 @@ def get_scansion(text, rhyme_analysis=False, rhythm_format="pattern"):
...
+        for liaison in (
+                ("synalepha",),
+                ("sinaeresis",),
+                ("sinaeresis", "synalepha"),
+                ("synalepha", "sinaeresis"),
+        ):
+            break_on_h = lambda liaison_type, syll1, syll2: (
+                    liaison_type == "synalepha"
+                    and syll2["syllable"][0].lower() == "h"
+            )
```
Comment (versae) [inline src/rantanplan/core.py]: Let's change it to:
```            def break_on_h(liaison_type, syllable_left, syllable_right):
                return (
                    liaison_type == "synalepha"
                    and syllable_right["syllable"][0].lower() == "h"
                )
```
And take it out of the for loops. We might take it to a different file if we add more breakage functions.

File: src/rantanplan/rhymes.py [hunk 6]
```diff
@@ -33,13 +34,18 @@
     (
         CONSONANT_RHYME,
         "sonnet",
-        r"(abba|abab|cddc|cdcd){2}((cd|ef){3}|(cde|efg){2})",
-        lambda lengths: all(13 > length > 10 for length in lengths)
+        r"(abba|abab|cddc|cdcd){2}((cd|ef){3}|(cde|efg){2}|[cde]{6})",
+        lambda lengths: all(13 > length > 9 for length in lengths)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Let's make it `all(14 > length > 9 for length in lengths)`

File: src/rantanplan/rhymes.py [hunk 7]
```diff
@@ -60,20 +66,28 @@
...
+                     for phonological_group in line["phonological_groups"]]
+        syllables_count = len(syllables)
+        syllables_stresses = [syllable["is_stressed"]
+                              for syllable in line["phonological_groups"]]
+        inverted_stresses = syllables_stresses[::-1]
+        last_stress_index = (
+            len(inverted_stresses) - inverted_stresses.index(True) - 1
+        )
+        ending = syllables[last_stress_index:]
+        endings.append((ending, syllables_count, last_stress_index - syllables_count))
```
Comment (versae) [inline src/rantanplan/rhymes.py]:
```
        endings.append(
            (ending, syllables_count, last_stress_index - syllables_count)
        )
```

File: src/rantanplan/rhymes.py [hunk 8]
```diff
@@ -204,14 +219,18 @@ def get_rhymes(stressed_endings, assonance=False, relaxation=False,
     codes, ending_codes, unrhymed_verses = get_clean_codes(
         stressed_endings, assonance, relaxation
     )
+    # print(codes, ending_codes, unrhymed_verses)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Remove

File: src/rantanplan/rhymes.py [hunk 9]
```diff
@@ -204,14 +219,18 @@ def get_rhymes(stressed_endings, assonance=False, relaxation=False,
...
     # Get the actual rhymes and endings adjusting for unrhymed verses
     rhyme_codes, endings = assign_letter_codes(
         codes, ending_codes, unrhymed_verses, offset
     )
+    # print(rhyme_codes, endings)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Remove

File: src/rantanplan/rhymes.py [hunk 10]
```diff
@@ -204,14 +219,18 @@ def get_rhymes(stressed_endings, assonance=False, relaxation=False,
...
     # Assign and reorder rhyme letters so first rhyme is always an 'a'
     rhymes = rhyme_codes_to_letters(rhyme_codes, unrhymed_verse_symbol)
+    # print(rhymes)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Remove

File: src/rantanplan/rhymes.py [hunk 11]
```diff
@@ -204,14 +219,18 @@ def get_rhymes(stressed_endings, assonance=False, relaxation=False,
...
     # Extract stress from endings
     stresses, unstressed_endings = split_stress(endings)
+    # print(rhymes, stresses, unstressed_endings)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Remove

File: src/rantanplan/rhymes.py [hunk 12]
```diff
@@ -220,6 +239,7 @@ def search_structure(rhyme, syllables_count, structure_key, structures=None):
     consonance, a rhyme pattern, and a condition on the lengths of sylalbles
     of lines. For the first matching structure, its index in STRUCTURES will
     be returned. An alternative STRUCTURES list can ba passed in structures."""
+    # print(rhyme, syllables_count, structure_key)
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Remove

File: tests/fixtures/couplet.json [hunk 13]
```diff
@@ -1 +1,208 @@
...
+      {
+        "syllable": "da",
+        "is_stressed": false
+      }
+    ],
+    "rhythm": {
+      "rhythmical_stress": "+--+--+--+-",
+      "type": "pattern"
+    },
+    "rhythmical_length": 11
```
Comment (versae) [inline tests/fixtures/couplet.json]: This is the all format of `rhythm` dict, isn't?

File: tests/test_rhymes.py [hunk 14]
```diff
@@ -62,28 +62,30 @@ def test_get_stressed_endings():
...
+         'rhythm': {'rhythmical_stress': '+-', 'type': 'pattern'},
+         'rhythmical_length': 2},
+        {'tokens': [{'word': [{'syllable': 'pri', 'is_stressed': False},
+                              {'syllable': 'sión', 'is_stressed': True}],
+                     'stress_position': -1}],
+         'phonological_groups': [{'syllable': 'pri', 'is_stressed': False},
+                                 {'syllable': 'sión',
+                                  'is_stressed': True}],
+         'rhythm': {'rhythmical_stress': '-+-', 'type': 'pattern'},
+         'rhythmical_length': 3}
```
Comment (versae) [inline tests/test_rhymes.py]: Shouldn't it be: `'rhythm': {'stress': '-+-', 'type': 'pattern', 'length': 3}`?

File: src/rantanplan/core.py [hunk 15]
```diff
@@ -40,74 +42,98 @@
...
+    ((?:(?:[bcdfghjklmnñpqstvy][hlr])|
+    (?:[bcdfghjklmnñpqrstvy][hr])|
+    (?:[bcdfghjklmnñpqrstvyz][h]))[aáeéiíoóuúü])|
     # 6: non-liquid consonant (adds hyphen)
     ([a-záéíóúñ][bcdfghjklmnñpqrstvxyz][aáeéiíoóuúüï])|
     # 7: vowel group (adds hyphen)
     ([aáeéíoóú][aáeéíoóú])|
     # 8: umlaut 'u' diphthongs
-    (ü[eií])|
+    (ü[iíaeo])|
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/core.py]: Yes, I didn't take into account all diphthongs, so I changed it

File: src/rantanplan/alternative_syllabification.py [hunk 16]
```diff
@@ -47,6 +51,7 @@
     "neodarwinista": "ne-o-dar-wi-nis-ta",
     "pajla": "paj-la",
     "penthouse": "pent-house",
+    "pirhuín": "pir-huín",
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/alternative_syllabification.py]: The alternative form It's already on the test

File: src/rantanplan/alternative_syllabification.py [hunk 17]
```diff
@@ -21,6 +21,10 @@
     "beethoveniano": "bee-tho-ve-nia-no",
     "byroniana": "by-ro-nia-na",
     "byroniano": "by-ro-nia-no",
+    "carhuacina": "car-hua-ci-na",
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/alternative_syllabification.py]: Done

File: src/rantanplan/core.py [hunk 18]
```diff
@@ -40,74 +42,98 @@
...
     # 9: Explicit hiatus with umlaut vowels, first part
     ([aeiou][äëïöü])|
     #10: Explicit hiatus with umlaut vowels, second part
-    ([äëïöü][a-z])|
+    ([üäëïö][a-z])|
```
[the remark this answers is not in the document]
Comment (alvp) [reply src/rantanplan/core.py]: I was trying a different rule to catch that and ended up reverting changes but putting the letters in different order

File: tests/fixtures/couplet.json (hunk 13 above)
Comment (alvp) [reply tests/fixtures/couplet.json]: Those fixtures need some more work actually. All of them have the enconding problem of converting to json, will solve it

File: tests/test_rhymes.py (hunk 14 above)
Comment (alvp) [reply tests/test_rhymes.py]: Same as above

File: src/rantanplan/rhymes.py [hunk 19]
```diff
@@ -52,6 +52,11 @@
         r"((.b)+)|((.a)+)",
         lambda lengths: statistics.median(lengths) == 8
     ), (
+        ASSONANT_RHYME,
+        "romance",
+        r"((.b)+)|((.a)+)",
+        lambda lengths: statistics.median(lengths) == 8
+     ), (
```
Comment (versae) [inline src/rantanplan/rhymes.py]: Duplicated code.

File: tests/fixtures/couplet.json (hunk 13 above)
Comment (versae) [reply tests/fixtures/couplet.json]: I fixed by using the one-liner JSON and writing directly to a file, not to a text output like Notebook.

**Pull request merged**